File: csharp/DiagramPrinter/DiagramPrinter.cs
using Microsoft.Extensions.Logging;

namespace DiagramPrinter;

public class FlowchartDiagramAdapter(FlowchartDiagram? diagram)
{
    public FlowchartDiagram? Diagram { get; } = diagram;

    public bool DiagramIsAvailable()
    {
        return Diagram != null;
    }
}

/**
 * This is a class you'd like to get under test so you can change it safely.
 */
public class DiagramPrinter
{
    public const string Spreadsheet = "Spreadsheet";
    public const string Pdf = "PDF";

    private readonly ILogger<DiagramPrinter> _logger = LoggingProvider.CreateLogger<DiagramPrinter>();
    
    public bool PrintSummary(FlowchartDiagram? diagram, string language, ref string summaryText)
    {
        return PrintSummary(new FlowchartDiagramAdapter(diagram), language, out summaryText);
    }

    /*
* Copy this checklist into the code and make a commit with message: [intention] - Discover Adapter Interface
* Identify every call to 'wrapper.Awkward'. Go through each one and decide which checklist to use. Either 'Discover Adapter Method' or 'Propagate Adapter'. Add the relevant checklist as a comment.
* Follow the new checklists for each call to 'wrapper.Awkward'. 
     */
    public static bool PrintSummary(FlowchartDiagramAdapter flowchartDiagramAdapter, string language, out string summaryText)
    {
        if (!flowchartDiagramAdapter.DiagramIsAvailable())
        {
            summaryText = "";
            return false;
        }

        // Use this checklist for code that uses wrapper.Awkward that you don't need to include in your unit tests.
        // * 'Extract method' for the code using 'wrapper.Awkward'. Preferably take the whole line of code into the new method. Make sure the wrapper class is one of the arguments.
        // * If the new method is not static, 'Introduce Parameter' for all the member variables it uses. Then make it static.
        // * 'Make method non static' and select the wrapper class as the instance to move it to.
        // * Delete unnecessary comments and make a commit [underway] - Discover Adapter Interface
        var summary = new DiagramSummary(language);
        summary.AddTitle(flowchartDiagramAdapter.Diagram.Name(), flowchartDiagramAdapter.Diagram.SerialNumber());
        summary.AddHeader(flowchartDiagramAdapter.Diagram.SummaryInformation());
        summary.AddImage(flowchartDiagramAdapter.Diagram.FlowchartThumbnail());
        summaryText = summary.Export();
        return true;
    }

    public bool PrintDiagram(FlowchartDiagram? diagram, string? folder = null, string? filename = null)
    {
        if (diagram == null)
        {
            return false;
        }

        var info = new DiagramMetadata(diagram);
        if (info.FileType == Pdf)
        {
            var targetFilename = GetTargetFilename(folder, filename);
            return diagram.FlowchartAsPdf().CopyFile(info.FullFilename, targetFilename, true);
        }

        if (info.FileType == Spreadsheet)
        {
            var targetFilename = GetTargetFilename(folder, filename);
            if (!targetFilename.EndsWith(".xls"))
                targetFilename += ".xls";
            return diagram.FlowchartDataAsSpreadsheet().CopyFile(info.FullFilename, targetFilename, true);
        }

        // default case - print to a physical printer
        return new DiagramPhysicalPrinter().DoPrint(diagram, info, GetTargetFilename(folder, filename));
    }

    public bool PrintReport(FlowchartDiagram? diagram, string reportTemplate, string? folder = null,
        string? filename = null, bool summarize = true)
    {
        if (diagram == null)
        {
            return false;
        }

        FlowchartReport report = diagram.Report();
        var targetFilename = GetTargetFilename(folder, filename);
        _logger.LogInformation(message: "Creating report for {name} to file {targetFilename}", diagram.Name(),
            targetFilename);

        if (summarize)
        {
            diagram = diagram.Summary();
            report.Close();
            report = diagram.Report();
            report.Open(true);
            _logger.LogInformation(message: "Switched to summary report for {name}", diagram.Name());
        }

        if (!report.isOpen())
        {
            _logger.LogError("Failed to open report for writing.");
            return false;
        }

        var data = diagram.ReportData();

        if (!ValidateReport(reportTemplate, data))
        {
            _logger.LogError("Failed to validate report template.");
            return false;
        }

        if (summarize)
        {
            data.Add(diagram.SummaryInformation());
            report.OpenWithContents(reportTemplate, data, true);
        }
        else
        {
            report.OpenWithContents(reportTemplate, data, false);
        }

        report.SaveToFile(targetFilename);
        _logger.LogInformation("Report creation succeeded");
        return true;
    }

    public bool PrintPages(FlowchartDiagram? diagram, PagesBuilder builder)
    {
        if (diagram == null)
        {
            return false;
        }

        FlowchartReportItems data = diagram.ReportData();
        List<DiagramPage> pages = diagram.PagesData();

        var report = new DiagramPagesReport();
        var reportPages = new List<DiagramReportPage>();

        foreach (var page in pages)
        {
            DiagramReportPage processedPage = builder.ProcessPage(page, data);
            reportPages.Add(processedPage);
        }

        return builder.Apply(report, reportPages);
    }


    private static string GetTargetFilename(string? folder, string? filename)
    {
        if (folder == null)
        {
            folder = Path.GetTempPath();
        }

        if (filename == null)
        {
            filename = Path.GetTempFileName();
        }

        var targetFilename = Path.Join(folder, filename);
        return targetFilename;
    }


    public bool ValidateReport(string template, FlowchartReportItems substitutions)
    {
        try
        {
            CreateReport(template, substitutions.ToArray());
            return true;
        }
        catch (System.FormatException e)
        {
            _logger.LogError("Report template did not match substitutions");
            return false;
        }
    }

    private string CreateReport(string template, object[] substitutions)
    {
        var report = string.Format(template, substitutions);
        return report;
    }
}

File: kotlin/src/main/kotlin/sammancoaching/DiagramPrinter.kt
package sammancoaching

import java.io.IOException
import java.nio.file.Paths
import java.util.*
import java.util.logging.Logger


/**
 * This is a class you'd like to get under test so you can change it safely.
 */
class DiagramPrinter {
    fun printSummary(diagram: FlowchartDiagram?, language: String?, summaryText: StringBuilder): Boolean {
        if (diagram == null) {
            summaryText.setLength(0) // Clear text
            return false
        }

        val summary: DiagramSummary = DiagramSummary(language)
        summary.addTitle(diagram.name, diagram.serialNumber)
        summary.addHeader(diagram.summaryInformation)
        summary.addImage(diagram.flowchartThumbnail)
        summaryText.append(summary.export())
        return true
    }

    @Throws(IOException::class)
    fun printDiagram(diagram: FlowchartDiagram?, folder: String?, filename: String?): Boolean {
        if (diagram == null) {
            return false
        }

        val info = DiagramMetadata(diagram)
        if (PDF.equals(info.fileType)) {
            val targetFilename = getTargetFilename(folder, filename)
            return diagram.flowchartAsPdf.copyFile(info.fullFilename, targetFilename, true)
        }

        if (SPREADSHEET.equals(info.fileType)) {
            var targetFilename = getTargetFilename(folder, filename)
            if (!targetFilename.endsWith(".xls")) {
                targetFilename += ".xls"
            }
            return diagram.flowchartDataAsSpreadsheet.copyFile(info.fullFilename, targetFilename, true)
        }

        // Default case - print to a physical printer
        return DiagramPhysicalPrinter().doPrint(diagram, info, getTargetFilename(folder, filename))
    }


    fun printReport(
        diagram: FlowchartDiagram?, reportTemplate: String, folder: String?,
        filename: String?, summarize: Boolean
    ): Boolean {
        var diagram: FlowchartDiagram? = diagram
        if (diagram == null) {
            return false
        }

        var report: FlowchartReport = diagram.report()
        val targetFilename = getTargetFilename(folder, filename)
        _logger.info(
            String.format(
                "Creating report for %s to file %s",
                *arrayOf<Any?>(diagram.name, targetFilename)
            )
        )

        if (summarize) {
            diagram = diagram.summary()
            report.close()
            report = diagram.report()
            report.open(true)
            _logger.info("Switched to summary report for " + diagram.name)
        }

        if (!report.isOpen()) {
            _logger.warning("Failed to open report for writing.")
            return false
        }

        val data: FlowchartReportItems = diagram.reportData()

        if (!validateReport(reportTemplate, data)) {
            _logger.warning("Failed to validate report template.")
            return false
        }

        if (summarize) {
            data.add(diagram.summaryInformation)
            report.openWithContents(reportTemplate, data, true)
        } else {
            report.openWithContents(reportTemplate, data, false)
        }

        report.saveToFile(targetFilename)
        _logger.info("Report creation succeeded")
        return true
    }

    fun printPages(diagram: FlowchartDiagram?, builder: PagesBuilder): Boolean {
        if (diagram == null) {
            return false
        }

        val data: FlowchartReportItems = diagram.reportData()
        val pages: MutableList<DiagramPage> = diagram.PagesData()

        val report = DiagramPagesReport()
        val reportPages: MutableList<DiagramReportPage?> = ArrayList<DiagramReportPage?>()

        for (page in pages) {
            val processedPage = builder.processPage(page, data)
            reportPages.add(processedPage)
        }

        return builder.apply(report, Collections.unmodifiableList<DiagramReportPage>(reportPages))
    }

    fun validateReport(template: String, substitutions: FlowchartReportItems): Boolean {
        try {
            createReport(template, substitutions.toArray())
            return true
        } catch (e: IllegalFormatException) {
            _logger.warning("Report template did not match substitutions")
            return false
        }
    }

    private fun createReport(template: String, substitutions: Array<Any>): String {
        val report = String.format(template, *substitutions)
        return report
    }

    companion object {
        const val SPREADSHEET: String = "Spreadsheet"
        const val PDF: String = "PDF"
        private val _logger: Logger = Logger.getLogger(DiagramPrinter::class.java.getName())

        private fun getTargetFilename(folder: String?, filename: String?): String {
            var folder = folder
            var filename = filename
            if (folder == null) {
                folder = System.getProperty("java.io.tmpdir") // Equivalent to Path.GetTempPath()
            }

            if (filename == null) {
                filename = Paths.get(folder, "tempFile").toString() // Simulate temp file creation
            }

            return Paths.get(folder, filename).toString()
        }
    }


}

